### PR TITLE
[Model Parallel] Add configure sharded model hook

### PR DESCRIPTION
**Pull request by kaushikb11**
## What does this PR do?

Adds a `configure sharded model` hook. This is required for both DeepSpeed and Fully Sharded. Both teams have defined a context in which you can wrap model layers to be sharded instantly.

#### Why does it need to be a hook?

It needs to be a hook because model layers can only be sharded instantly after torch distributed is setup. This ensures that the distributed environment is setup, then calls the hook. 

We also wrap the hook with the model parallel context for DeepSpeed/Fully Sharded so the user doesn't need to do anything extra.

#### What about fit/test/predict?

It will be treated like the `setup` hook for now, where we assume we're either using `fit/test/predict` and can call the hook for the user. We can iterate on this in further PRs.

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/accelerators/accelerator.py [hunk 1]
```diff
@@ -439,6 +440,18 @@ def results(self) -> Any:
         """
         return self.training_type_plugin.results
 
+    @contextlib.contextmanager
+    def model_parallel_context(self) -> Generator:
+        """
+        Provide hook to create modules in a parallel aware context. This is useful for when we'd like to
```
Comment (tchaton) [inline pytorch_lightning/accelerators/accelerator.py:446]:
```suggestion
        Provide hook to create modules in a distributed aware context. This is useful for when we'd like to
```

File: pytorch_lightning/core/hooks.py [hunk 2]
```diff
@@ -310,6 +310,13 @@ def on_post_move_to_device(self):
 
         """
 
+    def on_model_parallel_setup(self) -> None:
+        """
+        Hook to create modules in a parallel aware context. This is useful for when using sharded plugins,
```
Comment (tchaton) [inline pytorch_lightning/core/hooks.py:315]: same.

File: pytorch_lightning/trainer/trainer.py [hunk 3]
```diff
@@ -1084,6 +1085,15 @@ def call_setup_hook(self, model: LightningModule) -> None:
         self.setup(model, stage=state)
         model.setup(stage=state)
 
+    def call_model_parallel_hook(self, model: LightningModule) -> None:
+        # Call model parallel hook if accelerator requests. In some cases
+        # we will not call the hook; the hook has initialized the sharded model for example.
+        if self.accelerator.call_model_parallel_setup_hook:
+            self.on_model_parallel_setup(model)
+            with self.accelerator.model_parallel_context():
+                model.on_model_parallel_setup()
```
Comment (tchaton) [inline pytorch_lightning/trainer/trainer.py:1094]:
```suggestion
            with self.accelerator.model_parallel_context():
                self.on_model_parallel_setup(model)
                model.on_model_parallel_setup()
```

Comment (tchaton) [reply pytorch_lightning/trainer/trainer.py:1094]: I think the callback hook should also be within the context manager.

File: tests/accelerators/test_common.py [hunk 4]
```diff
@@ -44,3 +59,80 @@ def test_evaluate(tmpdir, trainer_kwargs):
     # make sure weights didn't change
     new_weights = model.layer_0.weight.clone().detach().cpu()
     torch.testing.assert_allclose(old_weights, new_weights)
+
+
+def test_model_parallel_setup_called(tmpdir):
+
+    class TestModel(BoringModel):
+
+        def on_model_parallel_setup(self):
```
Comment (tchaton) [inline tests/accelerators/test_common.py:68]: Add a check that the context manager is actually being called.

File: pytorch_lightning/accelerators/accelerator.py (hunk 1 above)
Comment (kaushikb11) [reply pytorch_lightning/accelerators/accelerator.py]: Done!

File: pytorch_lightning/core/hooks.py (hunk 2 above)
Comment (kaushikb11) [reply pytorch_lightning/core/hooks.py]: Done!

File: pytorch_lightning/trainer/trainer.py (hunk 3 above)
Comment (kaushikb11) [reply pytorch_lightning/trainer/trainer.py]: Good catch!

File: tests/accelerators/test_common.py (hunk 4 above)
Comment (kaushikb11) [reply tests/accelerators/test_common.py:68]: Thanks, it made me see an Issue in implementation. Fixinggg

**Comment by SeanNaren** [comment]
hey @kaushikb11 we might be able to move the hook call to be before the accelerator setup, so it happens within the environment: https://github.com/PyTorchLightning/pytorch-lightning/blob/feat/model_parallel_hook/pytorch_lightning/trainer/trainer.py#L440

```python
        # ----------------------------
        # SET UP TRAINING
        # ----------------------------
        self.call_hook("on_before_accelerator_backend_setup", model)
        self.accelerator.connect(model)
        self.accelerator.setup_environment()
        self.call_setup_hook(model)  # allow user to setup lightning_module in accelerator environment
        self.call_model_parallel_hook(model)  # allow user to setup in model parallel environment
        self.accelerator.setup(self, model)  # note: this sets up self.lightning_module
```

This would mean that we can wrap the model layers, and then set up the model/accelerator as required. I can verify this works with FSDP tomorrow!

File: pytorch_lightning/callbacks/base.py [hunk 5]
```diff
@@ -29,6 +29,9 @@ class Callback(abc.ABC):
     Subclass this class and override any of the relevant hooks
     """
 
+    def on_model_parallel_setup(self, trainer, pl_module: LightningModule) -> None:
+        """Called before model parallel accelerator setup"""
+
```
Comment (ananthsub) [inline pytorch_lightning/callbacks/base.py:34]: @SeanNaren n00b q: why is this needed as a callback hook too?

File: pytorch_lightning/core/hooks.py [hunk 6]
```diff
@@ -310,6 +310,13 @@ def on_post_move_to_device(self):
 
         """
 
+    def on_model_parallel_setup(self) -> None:
+        """
+        Hook to create modules in a distributed aware context. This is useful for when using sharded plugins,
+        where we'd like to shard the model instantly, which is useful for extremely large models
+        which can save memory and initialization time.
+        """
```
Comment (ananthsub) [inline pytorch_lightning/core/hooks.py:318]: this should specify whether it runs on fit/val/test/predict stages. otherwise it can be confusing why this is different from the existing `setup` hook

File: pytorch_lightning/callbacks/base.py (hunk 5 above)
Comment (SeanNaren) [reply pytorch_lightning/callbacks/base.py:34]: Allows us to do this:

For FSDP:
https://github.com/PyTorchLightning/pytorch-lightning/blob/feat/fsdp/pytorch_lightning/plugins/training_type/fully_sharded.py#L128-L147

For DeepSpeed:

https://github.com/PyTorchLightning/pytorch-lightning/blob/feat/ds_update/pytorch_lightning/plugins/training_type/deepspeed.py#L284

So for those really large models that we'd like to shard instantly, this provides a context manager to instantiate your layers. In FSDP, you'll have to call auto_wrap/wrap yourself, in DeepSpeed this is done for you (each module is automatically sharded, in FSDP we have a policy in place).

Reason it's done like this is because we do not have a distributed environment set up till the `Trainer` is called. then upon fit, we can setup the environment! Let me know if you do have ideas though!

File: pytorch_lightning/core/hooks.py (hunk 6 above)
Comment (SeanNaren) [reply pytorch_lightning/core/hooks.py:318]: Good idea, I'll update the docstring as to why this doesn't implement fit/val/test/predict stages

File: tests/accelerators/test_common.py (hunk 4 above)
Comment (tchaton) [reply tests/accelerators/test_common.py:73]: Might want to check the hook is not called within the callbacks.

Comment (kaushikb11) [reply tests/accelerators/test_common.py:73]: I think we can assume Users would be aware of the implications.

**Review by SeanNaren — APPROVED** [review]

File: pytorch_lightning/trainer/trainer.py [hunk 7]
```diff
@@ -436,6 +436,7 @@ def fit(
         self.accelerator.connect(model)
         self.accelerator.setup_environment()
         self.call_setup_hook(model)  # allow user to setup lightning_module in accelerator environment
+        self.call_model_parallel_hook(model)  # allow user to setup in model parallel environment
```
Comment (awaelchli) [inline pytorch_lightning/trainer/trainer.py:439]: why does it have to be a new hook. could it be combined with the setup hook?

Comment (SeanNaren) [reply pytorch_lightning/trainer/trainer.py:439]: I honestly don't know which is better, so please advise!

I added the hook because it makes it clearer to the user that this is specific to model parallelism; in this case model parallel plugins (Fully Sharded/DeepSpeed) will wrap this hook in a specific context allowing them to do model parallel as layers are initialised. This could also be done in the setup hook obviously, but makes it more elusive imo.

The other thing is there is some additional branching logic around the call. For DeepSpeed and FSDP we call it once within a training/testing/validation run even if there are a fit/test combo; this is because we want to instantiate the layers once in the model parallel context, and never again in the same session!

File: pytorch_lightning/accelerators/accelerator.py [hunk 8]
```diff
@@ -439,6 +440,18 @@ def results(self) -> Any:
         """
         return self.training_type_plugin.results
 
+    @contextlib.contextmanager
+    def model_parallel_context(self) -> Generator:
+        """
+        Provide hook to create modules in a distributed aware context. This is useful for when we'd like to
+        shard the model instantly, which is useful for extremely large models which can save memory and
+        initialization time.
```
Comment (carmocca) [inline pytorch_lightning/accelerators/accelerator.py:448]:
```suggestion
        shard the model instantly - useful for extremely large models. Can save memory and
        initialization time.
```

File: pytorch_lightning/accelerators/accelerator.py [hunk 9]
```diff
@@ -466,3 +479,27 @@ def connect_precision_plugin(self, plugin: PrecisionPlugin) -> None:
...
+        """
+        Allow model parallel hook to be called in suitable environments determined by the training type plugin.
+        This is useful for when we want to shard the model once within fit.
+        Returns: True if we want to call the model parallel setup hook.
+        """
+        return self.training_type_plugin.call_model_parallel_setup_hook
+
+    @call_model_parallel_setup_hook.setter
+    def call_model_parallel_setup_hook(self, mode: bool) -> None:
+        if isinstance(mode, bool):
```
Comment (carmocca) [inline pytorch_lightning/accelerators/accelerator.py:494]: do we have this `isinstance` for a reason? i think i'd yolo it and keep it simple. it's a setter

File: pytorch_lightning/callbacks/base.py (hunk 5 above)
Comment (carmocca) [reply pytorch_lightning/callbacks/base.py:34]: But those two snippets you shared don't use `on_model_parallel_setup`, they use `model_parallel_context`.

I think the question is more along the lines of "when would you override `on_model_parallel_setup`?"

File: pytorch_lightning/trainer/trainer.py (hunk 7 above)
Comment (carmocca) [reply pytorch_lightning/trainer/trainer.py:439]: -1 to merging them. The LightningModule/DataModulel setup hook isn't related at all with the `model_parallel_setup` hook. The usecase is different and the later doesn't get a `stage` argument.

> we call it once within a training/testing/validation run even if there are a fit/test combo

Although this is also the case for the other setups

Comment (SeanNaren) [reply pytorch_lightning/trainer/trainer.py:439]: @carmocca just to be clear, the `setup` hook is called multiple times just with different stages right? Whereas the model parallel setup hook is only called once within training or testing, and is never called again, but this depends on the plugin itself via the property: https://github.com/PyTorchLightning/pytorch-lightning/pull/6679/files#diff-667a2513c158c2b01d73b479e4dea75587d8531f9cd286891d34570a2fd145dcR1082

File: pytorch_lightning/callbacks/base.py (hunk 5 above)
Comment (SeanNaren) [reply pytorch_lightning/callbacks/base.py:34]: The short answer is when you'd like to shard your model instantly (to save initialization time or load a very large model).

Here is a hard case for DeepSpeed: https://github.com/SeanNaren/minGPT/blob/stage3/mingpt/model.py#L155-L161

We'll eventually write docs once the integrations are merged to explain the use cases in greater details.

File: pytorch_lightning/trainer/trainer.py (hunk 7 above)
Comment (carmocca) [reply pytorch_lightning/trainer/trainer.py:439]: if `setup` is called with no stage argument, it won't be called again regardless of training/testing. If it's called with a stage, then it won't be called again for that stage

File: pytorch_lightning/callbacks/base.py (hunk 5 above)
Comment (carmocca) [reply pytorch_lightning/callbacks/base.py:34]: wouldn't a name like `configure_sharded_model` or `configure_parallel_model` be more appropriate?

File: pytorch_lightning/trainer/trainer.py (hunk 7 above)
Comment (awaelchli) [reply pytorch_lightning/trainer/trainer.py:439]: @SeanNaren The docs explain that user can build their layers inside this hook.
Until now, if the user wanted to delay model construction, they did that in the setup hook, and the setup hook gets called on all accelerators/plugins, which made the code agnostic to that. Or is this not correct?

So if we want to avoid duplication, what is your advice? Let the user call setup() in the model_parallel_setup() hook?

Comment (SeanNaren) [reply pytorch_lightning/trainer/trainer.py:439]: Thanks @awaelchli, that's the issue I was trying to think through as well. I could imagine two version of the docs:

Obviously I'll add details explaining what things mean:

```
To build your models faster, or to fit a model that doesn't fit on a single machine but will fit on the world memory, you have to define your layers in the `setup` function as below.

class MyModel(pl.LightningModule):
    def setup(self, trainer, stage):
         self.linear = torch.nn.Linear(5, 5)

The setup function will be wrapped within a context manager that shards the modules as they are made. 
```

or

```
To build your models faster, or to fit a model that doesn't fit on a single machine but will fit on the world memory, you have to define your layers in the `configure_parallel_layers` function as below.

class MyModel(pl.LightningModule):
    def configure_parallel_layers(self, stage):
         self.linear = torch.nn.Linear(5, 5)

The setup function will be wrapped within a context manager that shards the modules as they are made.

If you've used the setup function to setup layers as suggested in the docs previous, then you can call the setup function directly in the code:

class MyModel(pl.LightningModule):
    def configure_parallel_layers(self, stage):
         self.setup(stage)
```

Which would you prefer @carmocca @awaelchli @tchaton

**Review by tchaton — APPROVED** [review]
LGTM !

File: pytorch_lightning/accelerators/accelerator.py [hunk 10]
```diff
@@ -439,6 +440,18 @@ def results(self) -> Any:
         """
         return self.training_type_plugin.results
 
+    @contextlib.contextmanager
+    def model_parallel_context(self) -> Generator:
```
Comment (carmocca) [inline pytorch_lightning/accelerators/accelerator.py:444]: Shouldn't this be `model_sharded_context`

since we went with `configure_sharded_model`

not `configure_parallel_model`

File: pytorch_lightning/trainer/trainer.py (hunk 7 above)
Comment (Borda) [reply pytorch_lightning/trainer/trainer.py]: both are interesting ways, It would be most important that I can create all with a single method if no space limitations and partial if there are some restrictions...
just to not diverse too much from mentioned docs I would slightly prefer calling setup inside parallel

**Review by carmocca — APPROVED** [review]

**Review by williamFalcon — APPROVED** [review]

File: pytorch_lightning/trainer/trainer.py [hunk 11]
```diff
@@ -1082,6 +1083,15 @@ def call_setup_hook(self, model: LightningModule) -> None:
...
 
+    def call_configure_sharded_model(self, model: LightningModule) -> None:
+        # Call configure sharded model hook if accelerator requests. In some cases
+        # we will not call the hook; the hook has initialized the sharded model for example.
+        if self.accelerator.call_configure_sharded_model_hook:
+            with self.accelerator.model_sharded_context():
+                model.configure_sharded_model()
+                self.configure_sharded_model(model)
+            self.accelerator.call_configure_sharded_model_hook = False
+
```
Comment (ananthsub) [inline pytorch_lightning/trainer/trainer.py:1094]: @SeanNaren @kaushikb11 why's this needed in the trainer? could this be part of the accelerator setup? in L440 instead?

Comment (SeanNaren) [reply pytorch_lightning/trainer/trainer.py:1094]: which part exactly? the trainer has control over the hook, we do not want the hook to be called within the accelerator i think

Comment (ananthsub) [reply pytorch_lightning/trainer/trainer.py:1094]: I see, on first read I thought that we could do this:
```
self.call_setup_hook(model)
self.accelerator.setup(self, model)  -> inside here apply the model sharded context and call model.configure_sharded_model()

# now that accelerator has finished doing the model sharding setup call the callbacks
self.configure_sharded_model(model)

but this splits the callback hook from the setup which can be less convenient

Comment (SeanNaren) [reply pytorch_lightning/trainer/trainer.py:1094]: Agree as long as we're ok with calling the model hook within the accelerator (cc @justusschock @awaelchli)
